osslsigncode.c

```diff
@@ -2571,27 +2624,18 @@ static int msi_verify_pkcs7(PKCS7 *p7, GsfInfile *infile, unsigned char *exdata,
 		if (exok) {
 			printf("\n");
 		} else {
-			tohex(exdata, hexbuf, EVP_MD_size(md));
-			printf("    MISMATCH!!! FILE HAS %s\n", hexbuf);
+			tohex(exdata, hexbuf, MIN((size_t)EVP_MD_size(md), exlen));
+			printf("    MISMATCH!!!\n\t\t\tFILE HAS : %s\n", hexbuf);
 		}
 	}
 #endif
-	printf("\n");
-	ret |= verify_pkcs7(p7, options);
-	printf("\n");
+	ret |= verify_pkcs7(p7, &allownest, options);
 	if (allownest) {
-		int has_sig = 0;
-		PKCS7 *p7nest = pkcs7_get_nested_signature(p7, &has_sig);
-		if (p7nest) {
-			int nest_ret = msi_verify_pkcs7(p7nest, infile, exdata, exlen, 0, options);
-			if (ret)
-				ret = nest_ret;
-			PKCS7_free(p7nest);
-		} else if (!p7nest && has_sig) {
-			printf("Failed to decode nested signature!\n");
-			ret = 1;
-		}
+		nest_ret = msi_verify_nested_signature(p7, infile, exdata, exlen, options);
+		if (ret)
+			ret = nest_ret;
```

Comment:
This causes a signature with an invalid hash but a valid nested signature to be accepted.  I guess this is not what you meant.